managed/services/supervisord/pmm_config.go

```diff
@@ -0,0 +1,219 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package supervisord provides facilities for working with Supervisord.
+package supervisord
+
+import (
+	"bytes"
+	"os"
+	"text/template"
+
+	"github.com/pkg/errors"
+)
+
+// SavePMMConfig renders and saves pmm config.
+func SavePMMConfig(params map[string]any) error {
+	cfg, err := marshalConfig(params)
+	if err != nil {
+		return err
+	}
+	if err := saveConfig(pmmConfig, cfg); err != nil {
+		return errors.Wrapf(err, "failed to save pmm config")
+	}
+	return nil
+}
+
+func marshalConfig(params map[string]any) ([]byte, error) {
+	var buf bytes.Buffer
+	if err := pmmTemplate.Execute(&buf, params); err != nil {
+		return nil, errors.Wrapf(err, "failed to render pmm template")
+	}
+	return buf.Bytes(), nil
+}
+
+// saveConfig saves config in default directory.
+func saveConfig(path string, cfg []byte) (err error) {
+	// read existing content
+	oldCfg, err := os.ReadFile(path) //nolint:gosec
+	if os.IsNotExist(err) {
+		err = nil
+	}
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// compare with new config
+	if bytes.Compare(cfg, oldCfg) == 0 {
+		// nothing to change
+		return nil
+	}
+
+	// restore old content
+	defer func() {
+		if err == nil {
+			return
+		}
+		if resErr := os.WriteFile(path, oldCfg, 0o644); resErr != nil { //nolint:gosec
+			err = errors.Wrapf(resErr, "failed to restore config: %s", err.Error())
+		}
+	}()
+
+	if err = os.WriteFile(path, cfg, 0o644); err != nil { //nolint:gosec
+		err = errors.Wrap(err, "failed to write new config")
+	}
+	return
+}
+
+var pmmTemplate = template.Must(template.New("").Option("missingkey=error").Parse(`
+[unix_http_server]
+chmod = 0700
+username = dummy
+password = dummy
+
+[supervisord]
+logfile = /srv/logs/supervisord.log
+logfile_maxbytes = 5MB
+logfile_backups = 1
+loglevel = info
+pidfile = /tmp/supervisord.pid
+nodaemon = true
+nocleanup = false
+user = root
+strip_ansi = false
+
+[supervisorctl]
+username = dummy
+password = dummy
+
+; we rewrite autostart to true during update or build.
+[program:pmm-update-perform-init]
+command = /usr/sbin/pmm-update -run-playbook -playbook=/usr/share/pmm-update/ansible/playbook/tasks/init.yml
+directory = /
+autorestart = unexpected
+priority=-1
+exitcodes = 0
+autostart = false
+startretries = 3
+startsecs = 1
+stopsignal = TERM
+stopwaitsecs = 300
+stdout_logfile = /srv/logs/pmm-update-perform-init.log
+stdout_logfile_maxbytes = 50MB
+stdout_logfile_backups = 3
+redirect_stderr = true
+{{ if not .DisableInternalDB}}
+[program:postgresql]
```

Comment:
Can we have empty lines between neighbouring sections (or `programs` in supervisor's terminology)?

Reply:
spaces between sections added